universe/Condition.cpp

```diff
@@ -9955,6 +9955,203 @@ unsigned int Not::GetCheckSum() const {
+            if (!matches.empty()) break;
+        }
+
+        // non matching items were already moved from matches to temp_non_matches
+        // move non matching items from temp_non_matches to non_matches
+        non_matches.insert(non_matches.end(),
+                           std::make_move_iterator(temp_non_matches.begin()),
+                           std::make_move_iterator(temp_non_matches.end()));
+    }
+}
+
+bool TopmostMatches::RootCandidateInvariant() const {
+    if (m_root_candidate_invariant != UNKNOWN_INVARIANCE)
+        return m_root_candidate_invariant == INVARIANT;
+
+    for (auto& operand : m_operands) {
+        if (!operand->RootCandidateInvariant()) {
+            m_root_candidate_invariant = VARIANT;
+            return false;
+        }
+    }
+    m_root_candidate_invariant = INVARIANT;
+    return true;
+}
+
+bool TopmostMatches::TargetInvariant() const {
+    if (m_target_invariant != UNKNOWN_INVARIANCE)
+        return m_target_invariant == INVARIANT;
+
+    for (auto& operand : m_operands) {
+        if (!operand->TargetInvariant()) {
+            m_target_invariant = VARIANT;
+            return false;
+        }
+    }
+    m_target_invariant = INVARIANT;
+    return true;
+}
+
+bool TopmostMatches::SourceInvariant() const {
+    if (m_source_invariant != UNKNOWN_INVARIANCE)
+        return m_source_invariant == INVARIANT;
+
+    for (auto& operand : m_operands) {
+        if (!operand->SourceInvariant()) {
+            m_source_invariant = VARIANT;
+            return false;
+        }
+    }
+    m_source_invariant = INVARIANT;
+    return true;
+}
+
+std::string TopmostMatches::Description(bool negated/* = false*/) const {
+    std::string values_str;
+    if (m_operands.size() == 1) {
+        values_str += UserString("DESC_TOPMOST_BEFORE_SINGLE_OPERAND");
+        // Pushing the negation to the enclosed conditions
+        values_str += m_operands[0]->Description(negated);
+        values_str += UserString("DESC_TOPMOST_AFTER_SINGLE_OPERAND");
+    } else {
+        values_str += UserString("DESC_TOPMOST_BEFORE_OPERANDS");
+        for (unsigned int i = 0; i < m_operands.size(); ++i) {
+            // Pushing the negation to the enclosed conditions
+            values_str += m_operands[i]->Description(negated);
+            if (i != m_operands.size() - 1) {
+                values_str += UserString("DESC_TOPMOST_BETWEEN_OPERANDS");
+            }
+        }
+        values_str += UserString("DESC_TOPMOST_AFTER_OPERANDS");
+    }
+    return values_str;
+}
+
+std::string TopmostMatches::Dump(unsigned short ntabs) const {
+    std::string retval = DumpIndent(ntabs) + "TopmostMatches [\n";
+    for (auto& operand : m_operands)
+        retval += operand->Dump(ntabs+1);
+    retval += DumpIndent(ntabs) + "]\n";
+    return retval;
+}
+
+void TopmostMatches::GetDefaultInitialCandidateObjects(const ScriptingContext& parent_context, ObjectSet& condition_non_targets) const {
+    // XXX unsure what this does
```

Comment:
need to understand if GetDefaultInitialCandidateObjects is correct, then remove comment

Reply:
`GetDefaultInitialCandidateObjects` narrows the search space of objects when no list of candidates is provided.

If a single object or a containers of matches and non-matches are passed to `Eval`, then just those objects are tested by the condition. But there is an `Eval` overload that just takes a set of matches, which it immediately clears, and then fills with any objects in the universe that the condition matches.

To make this faster, conditions have the `GetDefaultInitialCandidateObjects` function to initialize the search space of objects to test, rather than always testing all objects in the `Universe`. For example, `Source` initializes with just the `ScriptingContext`'s source object, and `Type` will initialize based on what object type is being matched (eg. ships for `OBJ_SHIP`).

For this case, it's acting like an `Or` condition, and thus shouldn't narrow the search space at all, so like `Or` you shouldn't overload `GetDefaultInitialCandidateObjects` and instead just use the base class behaviour.